File: day11-2.py
#!/usr/bin/env python3

import sys

def ints(itr):
    return [int(i) for i in itr]

with open(sys.argv[1], 'r') as f:
    file_lines = [l for l in f.read().strip().split('\n')]


in_nums = []

total = 0
result = 0
other = 0

grid = []
while True:
    for l in file_lines:
        vals = [int(x) for x in list(l)]
        grid += [vals]

        if False:
            total += 1

    break

print()
print('start')
for row in grid:
    print(''.join(map(str, row)))
print()
flashes=0
for step in range(int(1e10)):
    flashed = set()
    if step % 10 == 0:
        print(step)
    for row in grid:
        for x, cell in enumerate(row):
            row[x] += 1
    new_flashed = 0
    while True:
        incd = 0
        for y, row in enumerate(grid):
            for x, cell in enumerate(row):
                if cell > 9 and (x, y) not in flashed:
                    pts = [
                        (x, y-1),
                        (x, y+1),
                        (x-1, y),
                        (x+1, y),
                        (x-1, y-1),
                        (x+1, y+1),
                        (x-1, y+1),
                        (x+1, y-1),
                    ]
                    new_flashed += 1
                    for px, py in pts:
                        if px >= 0 and px < len(row) and py >= 0 and py < len(grid):
                            # print('upd', px, py)
                            grid[py][px] += 1
                            if grid[py][px] > 9:
                                incd += 1
                            flashed.add((x, y))
        if not incd:
            break
    flashes += new_flashed
    if new_flashed == len(grid) * len(grid[0]):
        print('ANS', step+1)
        break
    for x, y in flashed:
        grid[y][x] = 0
    # print()
    # print('step', step)
    # for row in grid:
    #     print(''.join(map(str, row)))
    # print()
print(flashes)

print(f'Total: {total}')
print(f'Result: {result}')
print(f'Other: {other}')

File: day11-2.cpp
#include "bits/stdc++.h"

using namespace std;

int main(int argc, char **argv) {
    freopen(argv[1], "r", stdin);

    vector<vector<int>> grid;
    while (!cin.eof()) {
        string line;
        cin >> line;
        if (line.length() != 0) {
            grid.emplace_back();
            auto& row = grid.back();

            for (char c : line) {
                row.push_back(c - '0');
            }
        }
    }

    int flashes = 0;
    vector<vector<bool>> flashed(grid.size());
    for (auto& row : flashed) {
        row.resize(grid[0].size());
    }

    for (int step = 0; step < 10000000; step++) {
        if (step % 10 == 0) {
            cout << step << '\n';
        }

        for (auto& row : grid) {
            for (auto& cell : row) {
                cell += 1;
            }
        }

        int new_flashes = 0;
        while (true) {
            int incd = 0;
            for (int y = 0; y < grid.size(); y++) {
                auto& row = grid[y];
                for (int x = 0; x < row.size(); x++) {
                    auto cell = row[x];
                    if (cell > 9 && !flashed[y][x]) {
                        new_flashes++;

                        for (int xoff = -1; xoff <= 1; xoff++) {
                            for (int yoff = -1; yoff <= 1; yoff++) {
                                if (xoff == 0 && yoff == 0) {
                                    continue;
                                }

                                auto px = x + xoff;
                                auto py = y + yoff;
                                if (px >= 0 && px < row.size() && py >= 0 && py < grid.size()) {
                                    grid[py][px] += 1;
                                    if (grid[py][px] > 9) {
                                        incd += 1;
                                    }
                                    flashed[y][x] = true;
                                }
                            }
                        }
                    }
                }
            }

            if (!incd) {
                break;
            }
        }

        flashes += new_flashes;
        if (new_flashes == grid.size() * grid[0].size()) {
            cout << "ANS " << step+1 << '\n';
            break;
        }

        for (auto& row : grid) {
            for (auto& cell : row) {
                if (cell > 9) {
                    cell = 0;
                }
            }
        }
        for (auto& row : flashed) {
            std::fill(row.begin(), row.end(), false);
        }
    }

    cout << flashes << '\n';
    return 0;
}
